ecllibrary/std/DataPatterns/Benford.ecl

```diff
@@ -0,0 +1,314 @@
+    // Critical value:  13.362  15.507  17.535  20.090  26.125
+    #UNIQUENAME(CHI_SQUARED_CRITICAL_VALUE_1);
+    #SET(CHI_SQUARED_CRITICAL_VALUE_1, 20.090); // 99% probability
+
+    // Chi-squared critical values for 9 degrees of freedom at various probabilities
+    // Probability:     0.90    0.95    0.975   0.99    0.999
+    // Critical value:  14.684  16.919  19.023  21.666  27.877
+    #UNIQUENAME(CHI_SQUARED_CRITICAL_VALUE_2);
+    #SET(CHI_SQUARED_CRITICAL_VALUE_2, 21.666); // 99% probability
+
+    #UNIQUENAME(CHI_SQUARED_CRITICAL_VALUE);
+    LOCAL %CHI_SQUARED_CRITICAL_VALUE% := IF(%clampedDigit% = 1, %CHI_SQUARED_CRITICAL_VALUE_1%, %CHI_SQUARED_CRITICAL_VALUE_2%);
+
+    #UNIQUENAME(expectedDistribution);
+    LOCAL %expectedDistribution% := DATASET
+        (
+            [
+                {1, -1, 30.1, 17.6, 12.5, 9.7, 7.9, 6.7, 5.8, 5.1, 4.6},
+                {2, 12.0, 11.4, 10.9, 10.4, 10.0, 9.7, 9.3, 9.0, 8.8, 8.5},
+                {3, 10.2, 10.1, 10.1, 10.1, 10.0, 10.0, 9.9, 9.9, 9.9, 9.8},
+                {4, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0}
+            ],
+            {
+                UNSIGNED1   pos,
+                DECIMAL4_1  zero,
+                DECIMAL4_1  one,
+                DECIMAL4_1  two,
+                DECIMAL4_1  three,
+                DECIMAL4_1  four,
+                DECIMAL4_1  five,
+                DECIMAL4_1  six,
+                DECIMAL4_1  seven,
+                DECIMAL4_1  eight,
+                DECIMAL4_1  nine
+            }
+        );
+
+    // Remove all spaces from field list so we can parse it more easily
+    #UNIQUENAME(trimmedFieldList);
+    LOCAL %trimmedFieldList% := TRIM((STRING)fieldListStr, ALL);
+
+    // Ungroup the given dataset, in case it was grouped
+    #UNIQUENAME(ungroupedInFile);
+    LOCAL %ungroupedInFile% := UNGROUP(inFile);
+
+    // Clamp the sample size to something reasonable
+    #UNIQUENAME(clampedSampleSize);
+    LOCAL %clampedSampleSize% := MAX(1, MIN(100, (INTEGER)sampleSize));
+
+    // Create a sample dataset if needed
+    #UNIQUENAME(sampledData);
+    LOCAL %sampledData% := IF
+        (
+            %clampedSampleSize% < 100,
+            ENTH(%ungroupedInFile%, %clampedSampleSize%, 100, 1, LOCAL),
+            %ungroupedInFile%
+        );
+
+    // Slim the dataset if the caller provided an explicit set of attributes;
+    // note that the TABLE function will fail if %trimmedFieldList% cites an
+    // attribute that is a child dataset (this is an ECL limitation)
+    #UNIQUENAME(workingInFile);
+    LOCAL %workingInFile% :=
+        #IF(%trimmedFieldList% = '')
+            %sampledData%
+        #ELSE
+            TABLE(%sampledData%, {#EXPAND(%trimmedFieldList%)})
+        #END;
+
+    // Helper function that returns the first non-zero digit in a string
+    #UNIQUENAME(FirstDigit);
+    LOCAL UNSIGNED1 %FirstDigit%(VARSTRING s, UNSIGNED1 pos) := EMBED(C++)
```

Comment:
Is the parameter a string or a varstring?   The casts in the calls are to varstrings, but the parameter is a string.  I suspect better to have a string parameter (len, ptr)

Reply:
Converted parameter to STRING.